operator/endpointgc/gc.go

```diff
@@ -1 +1,266 @@
+		switch owner.Kind {
+		case "Pod":
+			result := g.checkPodForCEP(resource.Key{Name: owner.Name, Namespace: cep.Namespace}, podStore, scopedLog)
+			if result.validated {
+				return result.shouldBeDeleted
+			}
+			podChecked = true
+		case "CiliumNode":
+			result := g.checkCiliumNodeForCEP(resource.Key{Name: owner.Name}, ciliumNodeStore, scopedLog)
+			if result.validated {
+				return result.shouldBeDeleted
+			}
+		}
+	}
+	if !podChecked {
+		// Check for a Pod in case none of the owners existed
+		// This keeps the old behavior even if OwnerReferences are missing
+		result := g.checkPodForCEP(
+			resource.Key{Name: cep.Name, Namespace: cep.Namespace}, podStore, scopedLog)
+		if result.validated {
+			return result.shouldBeDeleted
+		}
+	}
+	return true
+}
+
+func (g *GC) checkPodForCEP(key resource.Key, podStore resource.Store[*slim_corev1.Pod], scopedLog *logrus.Entry) deleteCheckResult {
+	pod, exists, err := podStore.GetByKey(key)
+	if err != nil {
+		scopedLog.WithError(err).Warn("Unable to get pod from store")
+	}
+	if !exists {
+		return deleteCheckResult{validated: false}
+	}
+	// In Kubernetes Jobs, Pods can be left in Kubernetes until the Job
+	// is deleted. If the Job is never deleted, Cilium will never receive a Pod
+	// delete event, causing the IP to be left in the ipcache.
+	// For this reason we should delete the ipcache entries whenever the pod
+	// status is either PodFailed or PodSucceeded as it means the IP address
+	// is no longer in use.
+	if k8sUtils.IsPodRunning(pod.Status) {
+		return deleteCheckResult{validated: true, shouldBeDeleted: false}
+	}
+	return deleteCheckResult{validated: true, shouldBeDeleted: true}
+}
+
+func (g *GC) checkCiliumNodeForCEP(key resource.Key, ciliumNodeStore resource.Store[*cilium_api_v2.CiliumNode], scopedLog *logrus.Entry) deleteCheckResult {
+	_, exists, err := ciliumNodeStore.GetByKey(key)
+	if err != nil {
+		scopedLog.WithError(err).Warn("Unable to get CiliumNode from store")
+	}
+	if !exists {
+		return deleteCheckResult{validated: false}
+	}
+	return deleteCheckResult{validated: true, shouldBeDeleted: false}
+}
+
+func (g *GC) deleteCEP(cep *cilium_api_v2.CiliumEndpoint, scopedLog *logrus.Entry, ctx context.Context) error {
+	ciliumClient := g.clientset.CiliumV2()
+	scopedLog = scopedLog.WithFields(logrus.Fields{
+		logfields.EndpointID: cep.Status.ID,
+	})
+	scopedLog.Debug("Orphaned CiliumEndpoint is being garbage collected")
+	propagationPolicy := meta_v1.DeletePropagationBackground // because these are const strings but the API wants pointers
+	err := ciliumClient.CiliumEndpoints(cep.Namespace).Delete(
+		ctx,
+		cep.Name,
+		meta_v1.DeleteOptions{
+			PropagationPolicy: &propagationPolicy,
+			// Set precondition to ensure we are only deleting CEPs owned by
+			// this agent.
+			Preconditions: &meta_v1.Preconditions{
+				UID: &cep.UID,
+			},
+		})
+	switch {
+	case err == nil:
+		metrics.EndpointGCObjects.WithLabelValues(metrics.LabelValueOutcomeSuccess).Inc()
```

Comment:
We recently introduced support for modular metrics in the operator https://github.com/cilium/cilium/pull/28005. We can now use `cell.Metric` to declare cell-specific metrics instead of using the legacy (global) ones. Do you mind modularize the metrics too?
You can find an example of how to do so [here](https://github.com/cilium/cilium/pull/28166/commits/62ef3f46d776b1203f7be1b0ab12d146d55e3f62) (CES controller metrics modularization)

Reply:
Done